### PR TITLE
[Merged by Bors] - feat(Algebra/Lie): Engel subalgebras

### PR DESCRIPTION
**Pull request by jcommelin**
Progress toward https://github.com/leanprover-community/mathlib4/issues/10065

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/LinearMap/End.lean [hunk 1]
```diff
@@ -203,6 +203,11 @@ theorem surjective_of_iterate_surjective {n : ℕ} (hn : n ≠ 0) (h : Surjectiv
   exact Surjective.of_comp h
 #align linear_map.surjective_of_iterate_surjective LinearMap.surjective_of_iterate_surjective
 
+lemma iterate_apply_eq_zero_of_le {m n : ℕ} {x : M} (hmn : m ≤ n) (hf : (f' ^ m) x = 0) :
+    (f' ^ n) x = 0 := by
+  obtain ⟨k, rfl⟩ := Nat.exists_eq_add_of_le hmn
+  rw [add_comm _ k, pow_add, LinearMap.mul_apply, hf, map_zero]
+
```
Comment (ocfnash) [inline Mathlib/Algebra/Module/LinearMap/End.lean:210]: This is exactly `LinearMap.pow_map_zero_of_le`.
```suggestion
```
I think the poor naming in this corner of the library contributed to not finding `LinearMap.pow_map_zero_of_le`. (There are a bunch of lemmas here that are about `pow` but whose name incorrectly says `iterate` and the lemma `LinearMap.pow_map_zero_of_le` should really be named `LinearMap.pow_apply_zero_of_le`.)

File: Mathlib/Algebra/Lie/OfAssociative.lean [hunk 2]
```diff
@@ -252,6 +252,63 @@ theorem LieSubalgebra.toEndomorphism_mk (K : LieSubalgebra R L) {x : L} (hx : x
...
+  | succ n ih => simp only [pow_succ', LinearMap.mul_apply, ih, LieSubmodule.coe_toEndomorphism]
+
+lemma LieSubalgebra.coe_ad (H : LieSubalgebra R L) (x y : H) :
+    (ad R H x y : L) = ad R L x y := rfl
+
+lemma LieSubalgebra.coe_ad_pow (H : LieSubalgebra R L) (x y : H) (n : ℕ) :
+    ((ad R H x ^ n) y : L) = (ad R L x ^ n) y := by
+  induction n generalizing y with
+  | zero => rfl
+  | succ n ih => simp only [pow_succ', LinearMap.mul_apply, ih, LieSubalgebra.coe_ad]
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/OfAssociative.lean:275]:
```suggestion
    ((ad R H x ^ n) y : L) = (ad R L x ^ n) y :=
  LieSubmodule.coe_toEndomorphism_pow R H L H.toLieSubmodule x y n
```

File: Mathlib/Algebra/Lie/OfAssociative.lean [hunk 3]
```diff
@@ -252,6 +252,63 @@ theorem LieSubalgebra.toEndomorphism_mk (K : LieSubalgebra R L) {x : L} (hx : x
...
+  | zero => simp
+  | succ n ih =>
+    rw [Finset.sum_antidiagonal_choose_succ_nsmul
+      (fun i j ↦ ⁅((ad R L x) ^ i) y, ((toEndomorphism R L M x) ^ j) z⁆) n]
+    simp only [pow_succ, LinearMap.mul_apply, ih, map_sum, map_nsmul,
+      toEndomorphism_lie, nsmul_add, sum_add_distrib]
+    convert add_comm _ _ using 4
+    rename_i ij hij
+    rw [mem_antidiagonal, add_comm] at hij
+    exact Nat.choose_symm_of_eq_add hij.symm
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/OfAssociative.lean:302]: I claim this is slightly less brittle / more maintainable:
```suggestion
    rw [add_comm]
    refine congr_arg₂ _ rfl (sum_congr rfl fun ⟨i, j⟩ hij ↦ ?_)
    replace hij : n = j + i := by rwa [mem_antidiagonal, add_comm, eq_comm] at hij
    rw [Nat.choose_symm_of_eq_add hij]
```
(though feel free to disagree)

File: Mathlib/Algebra/Lie/OfAssociative.lean [hunk 4]
```diff
@@ -252,6 +252,63 @@ theorem LieSubalgebra.toEndomorphism_mk (K : LieSubalgebra R L) {x : L} (hx : x
...
+    (ad R H x y : L) = ad R L x y := rfl
+
+lemma LieSubalgebra.coe_ad_pow (H : LieSubalgebra R L) (x y : H) (n : ℕ) :
+    ((ad R H x ^ n) y : L) = (ad R L x ^ n) y := by
+  induction n generalizing y with
+  | zero => rfl
+  | succ n ih => simp only [pow_succ', LinearMap.mul_apply, ih, LieSubalgebra.coe_ad]
+
+variable {L M}
+
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/OfAssociative.lean:278]:
```suggestion

local notation "φ" => LieModule.toEndomorphism R L M

```
I sometimes use this notation. It makes some of the below statements (and especially infoview state) a lot easier to read.

File: Mathlib/Algebra/Lie/EngelSubalgebra.lean [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  rcases hy with ⟨n, hn⟩
+  rw [mem_engel_iff]
+  use n+1
+  rw [pow_succ', LinearMap.mul_apply]
+  exact hn
+
+variable {R}
+
+open Filter in
+/-- A Lie-subalgebra of an artinian Lie algebra is self-normalizing
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/EngelSubalgebra.lean:91]: Some people disagree with me on this:
```suggestion
/-- A Lie-subalgebra of an Artinian Lie algebra is self-normalizing
```
so feel free to ignore. (I do note that we are capitalising `Lie` though.)

(IIRC Dijkstra joked that the supreme achievement is to have one's surname become an adjective that is so familiar that it is no longer capitalised.)

File: Mathlib/Algebra/Lie/EngelSubalgebra.lean [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    | succ k ih => simp only [pow_succ', LinearMap.mem_ker, LinearMap.mul_apply] at ih ⊢; apply ih
+  · rw [← Submodule.map_le_iff_le_comap]
+    apply le_sup_of_le_right
+    rw [Submodule.map_le_iff_le_comap]
+    rintro _ ⟨y, rfl⟩
+    simp only [pow_succ, LinearMap.mul_apply, Submodule.mem_comap, mem_coe_submodule]
+    apply aux₁
+    simp only [Submodule.coeSubtype, SetLike.coe_mem]
+
+/-- A Lie subalgebra of a noetherian Lie algebra is nilpotent
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/EngelSubalgebra.lean:133]:
```suggestion
/-- A Lie subalgebra of a Noetherian Lie algebra is nilpotent
```
(again, feel free to ignore)

**Review by ocfnash — APPROVED** [review]
Thanks, this is great.

Apart from the inline nitpicks, I wonder if we should take this as an opportunity to reorganise a few files. E.g., we could have a new `LieSubalgebra` folder containing:
`Basic` (current contents of `LieSubalgebra.lean`)
`Cartan`
`Engel`

bors d+

File: Mathlib/Algebra/Lie/EngelSubalgebra.lean [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+variable (R)
+
+/-- The Engel subalgebra `Engel R x` consists of
+all `y : L` such that `(ad R L x)^n` kills `y` for some `n`.
+
+Engel subalgebras are self-normalizing (`LieSubalgebra.normalizer_engel`),
+and minimal ones are nilpotent, hence Cartan subalgebras. -/
+@[simps!]
+def Engel (x : L) : LieSubalgebra R L :=
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/EngelSubalgebra.lean:49]: I _think_ initial uppercase is only for `Prop`-valued definitions.
```suggestion
def engel (x : L) : LieSubalgebra R L :=
```

File: Mathlib/Algebra/Module/LinearMap/End.lean (hunk 1 above)
Comment (jcommelin) [reply Mathlib/Algebra/Module/LinearMap/End.lean:210]: Ok, I'll just remove this. These changes should be done in a follow-up PR.

Comment (jcommelin) [inline Mathlib/Algebra/Module/LinearMap/End.lean:210]:
```suggestion
```

File: Mathlib/Algebra/Lie/EngelSubalgebra.lean (hunk 7 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/EngelSubalgebra.lean]: Thanks, fixed everywhere.

File: Mathlib/Algebra/Lie/EngelSubalgebra.lean (hunk 5 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/EngelSubalgebra.lean]: Thanks, fixed everywhere.

File: Mathlib/Algebra/Lie/EngelSubalgebra.lean (hunk 6 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/EngelSubalgebra.lean]: Idem

File: Mathlib/Algebra/Lie/OfAssociative.lean (hunk 2 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/OfAssociative.lean]: Thanks!

File: Mathlib/Algebra/Lie/OfAssociative.lean (hunk 4 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/OfAssociative.lean:276]: Thanks, implemented.

File: Mathlib/Algebra/Lie/OfAssociative.lean (hunk 3 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/OfAssociative.lean]: Thanks, I combined this with some other changes. Hopefully you like the improvement.

**Comment by jcommelin** [comment]
> Apart from the inline nitpicks, I wonder if we should take this as an opportunity to reorganise a few files. E.g., we could have a new `LieSubalgebra` folder containing: `Basic` (current contents of `LieSubalgebra.lean`) `Cartan` `Engel`

Sounds good. We can do that in a follow-up as well.

**Comment by jcommelin** [comment]
bors merge

**Comment by urkud** [comment]
Please run `lake exe shake --fix`
bors r-

**Pull request closed without merging**